File: coordinator/app/models/season.py
from datetime import datetime
from enum import StrEnum

from sqlalchemy import JSON, BigInteger, DateTime, Integer, String, func
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base


class SeasonStatus(StrEnum):
    PENDING = "pending"
    ACTIVE = "active"
    SETTLED = "settled"
    CANCELLED = "cancelled"


class ScoringMethod(StrEnum):
    PNL = "pnl"
    SHARPE = "sharpe"
    RISK_ADJUSTED = "risk_adjusted"


class Season(Base):
    __tablename__ = "seasons"

    id: Mapped[int] = mapped_column(primary_key=True)
    season_id_onchain: Mapped[int] = mapped_column(BigInteger, unique=True, index=True)
    season_pda: Mapped[str | None] = mapped_column(String(44), unique=True)
    authority: Mapped[str] = mapped_column(String(44), index=True)
    status: Mapped[SeasonStatus] = mapped_column(String(16), default=SeasonStatus.PENDING)
    name: Mapped[str] = mapped_column(String(64))
    description: Mapped[str] = mapped_column(String(256), default="")
    trading_universe: Mapped[list[str]] = mapped_column(JSON, default=list)
    max_agents: Mapped[int] = mapped_column(Integer)
    scoring_method: Mapped[ScoringMethod] = mapped_column(
        String(16), default=ScoringMethod.RISK_ADJUSTED
    )
    start_time: Mapped[int | None] = mapped_column(BigInteger, default=None)
    end_time: Mapped[int] = mapped_column(BigInteger)
    agent_count: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    onchain_synced_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), default=None
    )

File: programs/gladius/src/state/season.rs
use anchor_lang::prelude::*;

use crate::state::{
    SEASON_DESCRIPTION_MAX_LEN, SEASON_NAME_MAX_LEN, TRADING_UNIVERSE_MAX_LEN,
};

#[account]
#[derive(InitSpace, Debug)]
pub struct Season {
    pub season_id: u64,
    pub authority: Pubkey,
    pub status: SeasonStatus,
    pub config: SeasonConfig,
    /// `None` until `start_season` runs; `Some(unix_ts)` after.
    pub start_time: Option<i64>,
    pub end_time: i64,
    pub agent_count: u32,
    pub created_at: i64,
    pub bump: u8,
}

#[derive(
    AnchorSerialize, AnchorDeserialize, Clone, Copy, PartialEq, Eq, InitSpace, Debug,
)]
pub enum SeasonStatus {
    Pending,
    Active,
    Settled,
    Cancelled,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, InitSpace, Debug)]
pub struct SeasonConfig {
    #[max_len(SEASON_NAME_MAX_LEN)]
    pub name: String,
    #[max_len(SEASON_DESCRIPTION_MAX_LEN)]
    pub description: String,
    #[max_len(TRADING_UNIVERSE_MAX_LEN)]
    pub trading_universe: Vec<Pubkey>,
    pub max_agents: u32,
    pub scoring_method: ScoringMethod,
}

#[derive(
    AnchorSerialize, AnchorDeserialize, Clone, Copy, PartialEq, Eq, InitSpace, Debug,
)]
pub enum ScoringMethod {
    Pnl,
    Sharpe,
    RiskAdjusted,
}
